File: solution/398RandomPickIndex/398RandomPickIndex.cpp
#include <vector>
using namespace std;

class Solution {
private:
  unordered_map<int, int> m;
  vector<int> nums;

public:
  Solution(vector<int> nums) {
    this->nums = nums;
    for (int i = 0; i < nums.size(); i++) {
      m[nums[i]]++;
    }
    srand(nums.size());
  }

  int pick(int target) {
    int idx_m = rand() % m[target];
    int idx = 0;
    for (int i = 0; i < nums.size(); i++) {
      if (nums[i] == target) {
        if (idx == idx_m) {
          return i;
        } else {
          idx++;
        }
      }
    }
    return 0;
  }
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(nums);
 * int param_1 = obj.pick(target);
 */

int main() { return 0; }

File: src/398RandomPickIndex/398RandomPickIndex.rs
use rand::{thread_rng, Rng};

struct Solution {
    nums: Vec<i32>,
}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl Solution {

    fn new(nums: Vec<i32>) -> Self {
        Solution {
            nums,
        }
    }
    
    fn pick(&self, target: i32) -> i32 {
        let mut rng = thread_rng();
        let mut i: usize = 0;
        let mut index = 0;
        for k in 0..self.nums.len(){
            if self.nums[k] == target {
                let j: usize = rng.gen_range(0, i+1);
                if j == i {
                    index = k;
                }
                i += 1;
            }
        }
        
        index as i32
    }
}

/**
 * Your Solution object will be instantiated and called as such:
 * let obj = Solution::new(nums);
 * let ret_1: i32 = obj.pick(target);
 */
